library/psa_crypto.c

```diff
@@ -441,67 +441,48 @@ mbedtls_ecp_group_id mbedtls_ecc_group_of_psa( psa_ecc_family_t curve,
 }
 #endif /* defined(MBEDTLS_ECP_C) */
 
-static psa_status_t prepare_raw_data_slot( psa_key_type_t type,
-                                           size_t bits,
-                                           struct raw_data *raw )
+static psa_status_t validate_unstructured_key_bit_size( psa_key_type_t type,
+                                                        size_t bits )
 {
     /* Check that the bit size is acceptable for the key type */
     switch( type )
     {
         case PSA_KEY_TYPE_RAW_DATA:
-#if defined(MBEDTLS_MD_C)
```

Comment:
Conceptually:
a: I agree that code about algorithms or key types that are not supported downstream should be removed, but
b: how is this going to work for the accelerator case, which this PR is building towards?

The `MBEDTLS_XXX_C` defines are used to compile in/out support for a certain cryptographic algorithm, and will eat code space. For the accelerator case, you want input validation to happen in PSA Core, but the option to not have a software algorithm to back it, instead delegating to an accelerator. That means either validation and operations need to be split in two compilation units (with each their `_C` define?), or validation needs to happen inside PSA Core code.

For symmetric ciphers, validation is so very basic that I don't see why we'd require compiling in software support for the entire algorithm just to validate input. What if the core is being used for key storage at-rest only (e.g. in a TF-M model)? Are you going to require compiling in a full software ChaCha20 implementation just because the app storing the key is indicating it's a ChaCha20 key when storing it for later retrieval?

Reply:
`MBEDTLS_xxx_C` indicates that the feature is available to the application, regardless of how it's implemented (built-in code, `MBEDTLS_xxx_ALT`, or PSA driver).

Reply:
Though currently, the way you turn on/off features is by defining/undefining `MBEDTLS_xxx_C` macros, which means they're used to compile in software support as much as to indicate capabilities. You can't have `MBEDTLS_XXX_ALT` without also having `MBEDTLS_XXX_C`.

How do you see this getting resolved? If `MBEDTLS_XXX_C` takes on the meaning of "support available", then it should be defined through a combination of driver capability and software capability, and the software capability should be compiled in based on a different macro (to-be-introduced).
If `MBEDTLS_XXX_C` takes on the meaning of "software implementation compiled in" (which I think is the most common interpretation amongst users, given the prefix), then we should probably start defining `PSA_XXX_SUPPORTED` based on a combination of `MBEDTLS_XXX_C` macros and to-be-defined driver-provided macros.

Reply:
Currently, if you turn on `MBEDTLS_xxx_C` and `MBEDTLS_xxx_ALT`, the built-in implementation is not included in the build. Turning on `MBEDTLS_xxx_C` turns on the feature, and another option controls what implements the feature. It's important to preserve this property, both for backward compatibility and because it provides some independence between the application configuration (what cryptographic mechanisms do I need?) and the platform configuration (what cryptographic hardware is available?).

We need to gate the software implementation with some kind of `!MBEDTLS_PSA_xxx_DRIVER` in addition to `!MBEDTLS_xxx_ALT`. Names and granularity to be defined.

Reply:
Reverted through history rewrite.